### PR TITLE
mds: big refactor of MDS into MDSDaemon, MDSRank, MDSRank Dispatcher

### PR DESCRIPTION
**Pull request by jcsp**
This is a big interface cleanup:
 * Subsystems like server, mdcache etc now only see an MDSRank* instead of the whole MDS* caboodle.
 * The start/stop/dispatch hooks for MDSRank are in a separate MDSRankDIspatcher child class, that only MDSDaemon touches
 * The pre-taking-a-rank stuff (i.e. being in standby) behaviour all lives in MDSDaemon.

This gives us the nice property that nothing that requires an MDS rank is actually instantiated until we get a rank (i.e. construct MDSRank), and gives us a nice decrease in the visibility of parts that were effectively global before.

This does *not* do the separation of the lock (MDSRank still uses a reference to MDSDaemon::mds_lock) or dispatchers (MDSRank::ms_dispatch is called through from MDSDaemon).  Those will come along later -- I think it makes sense to let this stuff land before then so that we separate this many-LOC code movement from the more fine/fiddly concurrency parts.

### CODE REVIEW

**Comment by jcsp** [comment]
@gregsfortytwo passing unit tests now (it was the pool add/remove during tests that did it)

File: src/mds/MDSRank.h [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+    const int incarnation;
+
+  public:
+    mds_rank_t get_nodeid() const { return whoami; }
+    uint64_t get_metadata_pool();
+
+    // Reference to global MDS::mds_lock, so that users of MDSRank don't
+    // carry around references to the outer MDS, and we can substitute
+    // a separate lock here in future potentially.
+    Mutex &mds_lock;
```
Comment (gregsfortytwo) [inline src/mds/MDSRank.h]: I know it'd be a big line-count change, but I think maybe we should rename this to not be an alias. That would help keep the context we're working in clearer.
...or maybe we want them to alias while they're the same lock? Hrm.

**Comment by gregsfortytwo** [comment]
Regarding "mds: always delete MDS":
The original commit putting that in was to avoid assert failures on shutdown. Have you checked those no longer appear?

**Comment by gregsfortytwo** [comment]
Regarding "mds/StrayManager: don't acces OSDMap during construction":
We don't initialize max_purge_ops at all in the constructor anymore. Need to do that.

File: src/mds/MDSDaemon.cc [hunk 2]
```diff
@@ -0,0 +1,1318 @@
...
+      goto out;
+    }
+  }
+
+  if (whoami == MDS_RANK_NONE && (
+      new_state == MDSMap::STATE_STANDBY_REPLAY || new_state == MDSMap::STATE_ONESHOT_REPLAY)) {
+    whoami = mdsmap->get_mds_info_gid(mds_gid_t(monc->get_global_id())).standby_for_rank;
+  }
+
+  // mark down any failed peers
```
Comment (gregsfortytwo) [inline src/mds/MDSDaemon.cc]: Why did you move this block down? (It used to be above the standby machinations.)

File: src/mds/MDSDaemon.cc [hunk 3]
```diff
@@ -0,0 +1,1318 @@
...
+    dout(0) << "handle_mds_map mdsmap compatset " << mdsmap->compat
+	    << " not writeable with daemon features " << mdsmap_compat
+	    << ", killing myself" << dendl;
+    suicide();
+    goto out;
+  }
+
+  // If I was put into standby replay, but I am configured for a different standby
+  // type, ignore the map's state and request my standby type (only used
+  // for oneshot replay?)
```
Comment (gregsfortytwo) [inline src/mds/MDSDaemon.cc]: Yep! I don't remember exactly why we do this but I think it was necessary to get through the monitor startup dance successfully.

**Comment by gregsfortytwo** [comment]
Just a few thoughts; this looks really good. It's not merging on a minor conflict with the monitor optracker work; we can resolve at merge time or rebase as you like.

File: src/mds/MDSRank.h (hunk 1 above)
Comment (jcsp) [inline src/mds/MDSRank.h]: Yeah, the latter was my thinking -- it's less confusing (for us with the legacy code in our heads at least) if we keep calling mds_lock by its old name until it really is something different.

**Comment by jcsp** [comment]
> Regarding "mds/StrayManager: don't acces OSDMap during construction":
We don't initialize max_purge_ops at all in the constructor anymore. Need to do that.

Done

File: src/mds/MDSDaemon.cc (hunk 2 above)
Comment (jcsp) [inline src/mds/MDSDaemon.cc]: Uhm, I'm guessing I did that completely at random.  Moving it back up.

**Comment by jcsp** [comment]
> Regarding "mds: always delete MDS":
The original commit putting that in was to avoid assert failures on shutdown. Have you checked those no longer appear?

inasmuchas I haven't seen any asserts on shutdown... let's see if this makes it through an fs suite.

**Comment by dachary** [comment]
The bot failure is the usual jenkins bug when a rebase happens while the job is in progress, it can be ignored.

**Comment by jcsp** [comment]
@ukernel ok, I'm happy the failure was just the same valgrind error as on master, this should be good to merge
